OpenRiaServices.DomainServices.Client/Framework/Silverlight/Data/IndexedList.cs

```diff
@@ -0,0 +1,158 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Globalization;
+
+namespace OpenRiaServices.DomainServices.Client.Data
+{
+    /// <summary>
+    /// List with constant-time add/contains. Remove is O(n). Cannot store values
+    /// with duplicate hashes.
+    /// </summary>
+    internal class IndexedList : IList
+    {
+        public bool IsReadOnly => false;
+        public bool IsFixedSize => false;
+        public int Count => _list.Count;
+        public object SyncRoot => ((IList) _list).SyncRoot;
+        public bool IsSynchronized => ((IList) _list).IsSynchronized;
+
+        private const int NotContainedIndex = -1;
+
+        private readonly List<object> _list = new List<object>();
+
+        /// <summary>
+        /// Dictionary of { item, index in _list }
+        /// NOTE: this restricts usage to items with unique hashes
+        /// </summary>
+        private readonly Dictionary<object, int> _dict = new Dictionary<object, int>();
+
+        /// <summary>
+        /// Get/set value by index.
+        /// </summary>
+        public object this[int index]
+        {
+            get { return _list[index]; }
+            set
+            {
+                _list[index] = value;
+                _dict[value] = index;
```

Comment:
If updating the list we will incorrectly keep the old value in the dictionary, we should remove the previous value from _list[index] from the dictionary.
I believe the setter is unused so you could replace it with NotSupported if you like (and if you also don't see any code using the setter)